foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/descriptors/ClassDescriptor.java

```diff
@@ -3993,11 +4031,57 @@ public void postInitialize(AbstractSession session) throws DescriptorException {
 
         getCachePolicy().postInitialize(this, session);
 
+        postInitializeReturningPolicies();
+
         validateAfterInitialization(session);
 
         checkDatabase(session);
     }
 
+    private void postInitializeReturningPolicies() {
+        //Initialize ReturningPolicies
+        List<ReturningPolicy> returningPolicies = new ArrayList<>();
+        if (this.hasReturningPolicy()) {
+            returningPolicies.add(this.getReturningPolicy());
+        }
+        for (DatabaseMapping databaseMapping :this.getMappings()) {
+            ClassDescriptor referenceDescriptor = databaseMapping.getReferenceDescriptor();
```

Comment:
We only want the returning policies from the referenceDescriptor when the type is an AggregateObjectMapping otherwise we will get returning policies for related tables and not just the tables of this entity.

Reply:
Fixed, see line 4071